// @ramonornela/http/dist/umd/providers/http.js
(function (factory) {
    if (typeof module === 'object' && typeof module.exports === 'object') {
        var v = factory(require, exports); if (v !== undefined) module.exports = v;
    }
    else if (typeof define === 'function' && define.amd) {
        define(["require", "exports", '@angular/http', '@angular/core', './backend/xhr_backend', './exception', './plugins', '@ramonornela/configuration', '@ramonornela/url-resolver', './mapper', 'rxjs/Observable', 'rxjs/add/operator/toPromise', 'rxjs/add/observable/throw', 'rxjs/add/observable/defer', 'rxjs/add/operator/timeoutWith', 'rxjs/add/operator/map'], factory);
    }
})(function (require, exports) {
    "use strict";
    var http_1 = require('@angular/http');
    var core_1 = require('@angular/core');
    var xhr_backend_1 = require('./backend/xhr_backend');
    var exception_1 = require('./exception');
    var plugins_1 = require('./plugins');
    var configuration_1 = require('@ramonornela/configuration');
    var url_resolver_1 = require('@ramonornela/url-resolver');
    var mapper_1 = require('./mapper');
    var Observable_1 = require('rxjs/Observable');
    require('rxjs/add/operator/toPromise');
    require('rxjs/add/observable/throw');
    require('rxjs/add/observable/defer');
    require('rxjs/add/operator/timeoutWith');
    require('rxjs/add/operator/map');
    exports.RequestDefaultOptionsToken = new core_1.OpaqueToken('REQUESTDEFAULTOPTIONSTOKEN');
    exports.DefaultOptionsToken = new core_1.OpaqueToken('DEFAULTOPTIONSTOKEN');
    var KEY_CONFIG = 'http';
    var Http = (function () {
        function Http(http, events, plugins, config, requestFactory, defaultOptionsRequest, defaultOptions) {
            this.http = http;
            this.events = events;
            this.plugins = plugins;
            this.requestFactory = requestFactory;
            this.options = {};
            this.requestOptions = {};
            this.lastRequest = null;
            this.requests = {};
            this.runEvent('preRequest');
            this.runEvent('postRequest');
            this.runEvent('postRequestSuccess');
            this.runEvent('postRequestError');
            if (config) {
                var httpConfig = config.get(KEY_CONFIG) || {};
                if (!defaultOptionsRequest && httpConfig.defaultOptionsRequest) {
                    defaultOptionsRequest = httpConfig.defaultOptionsRequest;
                }
                if (!defaultOptions && httpConfig.defaultOptions) {
                    defaultOptions = httpConfig.defaultOptions;
                }
            }
            if (defaultOptionsRequest) {
                this.setDefaultRequestOptions(defaultOptionsRequest);
            }
            if (defaultOptions) {
                this.setDefaultOptions(defaultOptions);
            }
        }
        Http.prototype.getRequestFactory = function () {
            return this.requestFactory;
        };
        Http.prototype.throwsException = function (throws) {
            this.plugins.setThrowsException(throws);
            return this;
        };
        Http.prototype.canRetry = function (id) {
            if (id) {
                return this.requests[id] !== undefined;
            }
            return this.lastRequest !== null;
        };
        Http.prototype.getLastRequest = function () {
            return this.lastRequest;
        };
        Http.prototype.retryRequest = function (id) {
            if (id) {
                if (!this.requests[id]) {
                    throw new Error(id + " not exists to retry");
                }
                return this.request(this.requests[id]);
            }
            return this.request(this.lastRequest);
        };
        Http.prototype.request = function (url, params, requestOptions, options) {
            var _this = this;
            if (!(url instanceof url_resolver_1.Request) && arguments.length === 1 && typeof url === 'object') {
                var objParams = url;
                url = objParams.url;
                params = objParams.params;
                requestOptions = objParams.requestOptions;
                options = objParams.options;
            }
            if (requestOptions && this.checkForOptions(requestOptions)) {
                options = requestOptions;
                requestOptions = null;
            }
            options = options || {};
            this.applyDefaultOptions(options);
            this.lastRequest = null;
            if (options.retry) {
                this.lastRequest = {
                    url: url,
                    params: params,
                    requestOptions: requestOptions,
                    options: options
                };
                this.requests[url] = this.lastRequest;
            }
            requestOptions = requestOptions || {};
            requestOptions = Object.assign({}, this.requestOptions, requestOptions);
            if (typeof url === 'string' && this.requestFactory) {
                if (this.requestFactory.getMetadata().has(url)) {
                    url = this.requestFactory.create(url, params, requestOptions);
                    requestOptions = null;
                }
            }
            if (options.pluginsOptions) {
                this.plugins.setOptions(options.pluginsOptions);
            }
            var responseObservable = this.http.request(url, requestOptions);
            if (options.timeout) {
                responseObservable = responseObservable
                    .timeoutWith(options.timeout, Observable_1.Observable.defer(function () {
                    var err = new exception_1.TimeoutException();
                    _this.events.publish(xhr_backend_1.HttpEvents.POST_REQUEST_ERROR, err);
                    return Observable_1.Observable.throw(err);
                }));
            }
            if (options.mapper instanceof mapper_1.Mapper) {
                responseObservable = responseObservable.map(function (resp) { return options.mapper.transform(resp); });
            }
            return responseObservable;
        };
        Http.prototype.requestPromise = function (url, params, requestOptions, options) {
            return this.request.apply(this, arguments).toPromise();
        };
        Http.prototype.get = function (url, params, requestOptions, options) {
            requestOptions = requestOptions || {};
            requestOptions.method = 'GET';
            return this.request(url, params, requestOptions, options);
        };
        Http.prototype.post = function (url, params, requestOptions, options) {
            requestOptions = requestOptions || {};
            requestOptions.method = 'POST';
            return this.request(url, params, requestOptions, options);
        };
        Http.prototype.put = function (url, params, requestOptions, options) {
            requestOptions = requestOptions || {};
            requestOptions.method = 'PUT';
            return this.request(url, params, requestOptions, options);
        };
        Http.prototype.delete = function (url, params, requestOptions, options) {
            requestOptions = requestOptions || {};
            requestOptions.method = 'DELETE';
            return this.request(url, params, requestOptions, options);
        };
        Http.prototype.patch = function (url, params, requestOptions, options) {
            requestOptions = requestOptions || {};
            requestOptions.method = 'PATCH';
            return this.request(url, params, requestOptions, options);
        };
        Http.prototype.head = function (url, params, requestOptions, options) {
            requestOptions = requestOptions || {};
            requestOptions.method = 'HEAD';
            return this.request(url, params, requestOptions, options);
        };
        Http.prototype.checkForOptions = function (obj) {
            var properties = ['mapper', 'timeout', 'retry', 'pluginsOptions'];
            for (var _i = 0, properties_1 = properties; _i < properties_1.length; _i++) {
                var prop = properties_1[_i];
                if (prop in obj) {
                    return true;
                }
            }
            return false;
        };
        Http.prototype.applyDefaultOptions = function (options) {
            options.mapper = options.mapper || this.options.mapper;
            options.timeout = options.timeout || this.options.timeout;
            options.pluginsOptions = options.pluginsOptions || this.options.pluginsOptions;
            options.retry = options.retry || this.options.retry;
        };
        Http.prototype.setDefaultOptions = function (options) {
            this.options = options;
            return this;
        };
        Http.prototype.setTimeout = function (timeout) {
            this.options.timeout = timeout;
            return this;
        };
        Http.prototype.setMapper = function (mapper) {
            this.options.mapper = mapper;
            return this;
        };
        Http.prototype.setRetry = function (retry) {
            this.options.retry = retry;
            return this;
        };
        Http.prototype.setPluginsOptions = function (options) {
            this.options.pluginsOptions = options;
            return this;
        };
        Http.prototype.getPlugins = function () {
            return this.plugins;
        };
        Http.prototype.getEvents = function () {
            return this.events;
        };
        Http.prototype.getPlugin = function (name) {
            return this.getPlugins().get(name);
        };
        Http.prototype.setDefaultRequestOptions = function (options) {
            if (this.requestFactory) {
                this.requestFactory.setDefaultOptions(options);
                return this;
            }
            this.requestOptions = options;
            return this;
        };
        Http.prototype.runEvent = function (method) {
            var _this = this;
            var methodName = [
                'on',
                method.charAt(0).toUpperCase(),
                method.slice(1)
            ].join('');
            this.events[methodName].call(this.events, function (req) {
                _this.plugins.runEvent(method, [req]);
                if (method === 'postRequest' || method === 'postRequestError') {
                    _this.plugins.cleanOptions();
                }
            });
        };
        Http.decorators = [
            { type: core_1.Injectable },
        ];
        Http.ctorParameters = [
            { type: http_1.Http, },
            { type: xhr_backend_1.HttpEvents, },
            { type: plugins_1.Plugins, },
            { type: configuration_1.Config, decorators: [{ type: core_1.Optional },] },
            { type: url_resolver_1.Request, decorators: [{ type: core_1.Optional },] },
            { type: undefined, decorators: [{ type: core_1.Optional }, { type: core_1.Inject, args: [exports.RequestDefaultOptionsToken,] },] },
            { type: undefined, decorators: [{ type: core_1.Optional }, { type: core_1.Inject, args: [exports.DefaultOptionsToken,] },] },
        ];
        return Http;
    }());
    exports.Http = Http;
});
